### PR TITLE
osc-staging: provide rebuild_check() to conditionally rebuild failures.

### PR DESCRIPTION
**Pull request by jberry-suse**
Further the automation cause!

### CODE REVIEW

**Comment by jberry-suse** [comment]
Presumably, the check is simpler than I first thought. Each time a staging is accepted every package must have built successfully. That means there is no need to figure out when the staging is started as there should never be more than 3 failures after the last success (at least for auto rebuild).

The log check simply looks for stalls indicated by:
```
Job seems to be stuck here, killed.
```

**Comment by jberry-suse** [comment]
openSUSE:Factory:Staging:D had a kernel build that got stuck and this code properly detected the stuck state (I commented out the fail_count check. :)

I added this to my automatic set of commands running against Leap 42.3 (up to 4 now).

**Comment by jberry-suse** [comment]
Running this `rebuild` command on Factory resulted in an ever increasing number of `skipped` until all were in that category. This seems to work quite well.

After a few runs:

```bash
$ osc staging rebuild
rebuild openSUSE:Factory:Staging:A/yast2-ruby-bindings/standard/i586 skipped
rebuild openSUSE:Factory:Staging:A/yast2-ruby-bindings/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:A/yast2-ruby-bindings/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:A/libmicrohttpd/standard/i586 skipped
rebuild openSUSE:Factory:Staging:B/pulseaudio/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B:DVD/gdm/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B/man/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B:DVD/emacs/standard/ppc64le ok
rebuild openSUSE:Factory:Staging:B/libmicrohttpd/standard/i586 skipped
rebuild openSUSE:Factory:Staging:B/syslogd/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B/syslogd/standard/i586 skipped
rebuild openSUSE:Factory:Staging:B:DVD/sanlock/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B/man/standard/i586 skipped
rebuild openSUSE:Factory:Staging:B/build/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:B/build/standard/i586 skipped
rebuild openSUSE:Factory:Staging:B/man/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:B:DVD/sanlock/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:B/syslogd/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:B/build/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B:DVD/i4l-base/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:B/libmicrohttpd/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:G:DVD/deja-dup/standard/x86_64 skipped
rebuild openSUSE:Factory:Staging:G:DVD/deja-dup/standard/ppc64le skipped
rebuild openSUSE:Factory:Staging:J:DVD/Test-DVD-ppc64le/images/x86_64 skipped
```

**Comment by jberry-suse** [comment]
I noticed a few more stuck packages in Factory. It seems to occur more often than in Leap, probably since more packages are update and thus built so there is more opportunity to get stuck.

File: osclib/stagingapi.py [hunk 1]
```diff
@@ -736,22 +738,82 @@ def check_ring_packages(self, project, requests):
...
+                else:
+                    break
+        return fail_count
+
+    # Modfied from osc.core.print_jobhistory()
+    def job_history_get(self, project, repository, architecture, package=None):
+        query = {}
+        if package:
+            query['package'] = package
+        u = makeurl(self.apiurl, ['build', project, repository, architecture, '_jobhistory'], query)
```
Comment (lnussel) [inline osclib/stagingapi.py]: I think there should be a limit here. History could be quite long.

**Comment by jberry-suse** [comment]
Added limit with default to 20.

**Comment by DimStar77** [comment]
I'm basically fine with auto-triggering 'stuck' builds - but we should keep on having a feeling which packages this impacts. If it's always the same ones, it's questionable at best if this is really an OBS issue or a package issue.

auto-retriggering might take this perceptive feeling away from us

**Comment by jberry-suse** [comment]
Could keep track of which packages are re-triggered. Either parse the logs from where this runs as a service or the rebuild information on OBS side.

**Pull request merged**